simpleclient/src/main/java/io/prometheus/client/TimeWindowQuantiles.java

```diff
@@ -1,54 +1,99 @@
+    ConcurrentDoubleHistogram[] emptyBuckets = new ConcurrentDoubleHistogram[ageBuckets];
     for (int i = 0; i < ageBuckets; i++) {
-      this.ringBuffer[i] = new CKMSQuantiles(quantiles);
+      emptyBuckets[i] = createBucket();
     }
-    this.currentBucket = 0;
-    this.lastRotateTimestampMillis = System.currentTimeMillis();
-    this.durationBetweenRotatesMillis = TimeUnit.SECONDS.toMillis(maxAgeSeconds) / ageBuckets;
+    this.buckets = new AtomicReference<ConcurrentDoubleHistogram[]>(emptyBuckets);
+    this.lastRotateTimestampNanos = new AtomicLong(System.nanoTime());
+    this.durationBetweenRotatesNanos = TimeUnit.SECONDS.toNanos(maxAgeSeconds) / ageBuckets;
+  }
+
+  private ConcurrentDoubleHistogram createBucket() {
+    ConcurrentDoubleHistogram bucket = new ConcurrentDoubleHistogram(highestToLowestValueRatio, numberOfSignificantValueDigits);
+    bucket.setAutoResize(true);
+
+    return bucket;
+  }
+
+  public double get(double quantile) {
+    // On concurrent `get` and `rotate`, it is acceptable to `get` the sample from an outdated `bucket`.
+    ConcurrentDoubleHistogram currentBucket = getCurrentBucket();
+    return currentBucket.getTotalCount() == 0 ? Double.NaN : currentBucket.getValueAtPercentile(quantile * 100.0);
+  }
+
+  public double getMin() {
+    ConcurrentDoubleHistogram currentBucket = getCurrentBucket();
+    return currentBucket.getTotalCount() == 0 ? Double.NaN : currentBucket.getMinValue();
   }
 
-  public synchronized double get(double q) {
-    CKMSQuantiles currentBucket = rotate();
-    return currentBucket.get(q);
+  public double getMax() {
+    ConcurrentDoubleHistogram currentBucket = getCurrentBucket();
+    return currentBucket.getTotalCount() == 0 ? Double.NaN : currentBucket.getMaxValue();
   }
 
-  public synchronized void insert(double value) {
+  public void insert(double value) {
+    // On concurrent `insert` and `rotate`, it should be acceptable to lose the measurement in the newest `bucket`.
     rotate();
-    for (CKMSQuantiles ckmsQuantiles : ringBuffer) {
-      ckmsQuantiles.insert(value);
+
+    for (ConcurrentDoubleHistogram bucket : buckets.get()) {
+      bucket.recordValue(value);
     }
   }
 
-  private CKMSQuantiles rotate() {
-    long timeSinceLastRotateMillis = System.currentTimeMillis() - lastRotateTimestampMillis;
-    while (timeSinceLastRotateMillis > durationBetweenRotatesMillis) {
-      ringBuffer[currentBucket] = new CKMSQuantiles(quantiles);
-      if (++currentBucket >= ringBuffer.length) {
-        currentBucket = 0;
+  private ConcurrentDoubleHistogram getCurrentBucket() {
+    rotate();
+
+    return buckets.get()[0]; // oldest bucket
+  }
+
+  private void rotate() {
+    // On concurrent `rotate` and `rotate`:
+    //  - `currentTime` is cached to reduce thread contention.
+    //  - `lastRotate` is used to ensure the correct number of rotations.
+
+    // Correctness is guaranteed by `volatile` memory access ordering and visibility semantics.
```

Comment:
I don't see any use of volatile here. Are you sure this is right?

Reply:
Volatile inside atomic values was meant. I changed it to `atomic`.